libimage/runtime_test.go

```diff
@@ -67,41 +67,32 @@ func testRuntimePullImage(t *testing.T, r *Runtime, ctx context.Context, imageNa
 
 func TestTmpdir(t *testing.T) {
 	tmpStr := "TMPDIR"
-	tmp, tmpSet := os.LookupEnv(tmpStr)
-
 	confStr := "CONTAINERS_CONF"
-	conf, confSet := os.LookupEnv(confStr)
-
-	os.Setenv(confStr, "testdata/containers.conf")
-	_, err := config.Reload()
-	require.NoError(t, err)
 
-	tmpd, err := tmpdir()
-	require.NoError(t, err)
-	require.Equal(t, "/tmp/from/containers.conf", tmpd)
+	t.Run(confStr, func(t *testing.T) {
+		t.Setenv(confStr, "testdata/containers.conf")
+		_, err := config.Reload()
+		require.NoError(t, err)
 
-	if confSet {
-		os.Setenv(confStr, conf)
-	} else {
-		os.Unsetenv(confStr)
-	}
-	_, err = config.Reload()
-	require.NoError(t, err)
+		tmpd, err := tmpdir()
+		require.NoError(t, err)
+		require.Equal(t, "/tmp/from/containers.conf", tmpd)
+	})
 
-	os.Unsetenv(tmpStr)
-	tmpd, err = tmpdir()
-	require.NoError(t, err)
-	require.Equal(t, "/var/tmp", tmpd)
+	t.Run(tmpStr, func(t *testing.T) {
+		t.Setenv(tmpStr, "/tmp/test")
+		tmpd, err := tmpdir()
+		require.NoError(t, err)
+		require.Equal(t, "/tmp/test", tmpd)
 
-	os.Setenv(tmpStr, "/tmp/test")
-	tmpd, err = tmpdir()
-	require.NoError(t, err)
-	require.Equal(t, "/tmp/test", tmpd)
-	if tmpSet {
-		os.Setenv(tmpStr, tmp)
-	} else {
 		os.Unsetenv(tmpStr)
-	}
+		_, err = config.Reload()
```

Comment:
If I understand the code correctly, `config` is actually affected by `confStr`, not by `tmpStr`. So, the reload “needs to” happen when we change `confStr` == when we return from the first nested test; it doesn’t really need to happen for `tmpStr` changes.

Non-blocking: It would be a bit less dependent on the underlying implementation to trigger a reload before  _each of the 3_ `tmpdir()` calls.

Reply:
You are right, I think I fixed this properly now by using a t.Cleanup() to reload the config after the env has been reset. This ensures that even on a test failure above we reset the conf correctly and not cause weird cascading failures in other tests.